### PR TITLE
Snapshot ringbuffer implementation

### PR DESCRIPTION
**Pull request by z-j-lin**
## Scope

What is changing with this PR?
snapshot storage
## Why?
implements ringbuffer to store snapshots
Why are we doing this?
save gas 
## Todos
If any, what are the follow-up tasks required other than merging this PR?  Have they been arranged?
- [ ] ???

### CODE REVIEW

File: bridge/contracts/libraries/snapshots/SnapshotsStorage.sol [hunk 1]
```diff
@@ -4,13 +4,14 @@ pragma solidity ^0.8.11;
...
+import "contracts/libraries/snapshots/SnapshotRingBuffer.sol";
 
-abstract contract SnapshotsStorage is ImmutableETHDKG, ImmutableValidatorPool {
+abstract contract SnapshotsStorage is ImmutableETHDKG, ImmutableValidatorPool, SnapshotRingBuffer {
     uint256 internal immutable _epochLength;
 
     uint256 internal immutable _chainId;
 
-    uint32 internal _epoch;
+    // uint32 internal _epoch;
```
Comment (diligentis) [inline bridge/contracts/libraries/snapshots/SnapshotsStorage.sol:14]: Delete commented out code.

File: bridge/test/contract-mocks/snapshotRingBuffer/snapshotStorageMock.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    bytes32 internal constant SNAPSHOT_INCORRECT_BLOCK_HEIGHT = "406"; //"Snapshots: Incorrect AliceNet height for snapshot!"
+    bytes32 internal constant SNAPSHOT_INCORRECT_CHAIN_ID = "407"; //"Snapshots: Incorrect chainID for snapshot!"
+    bytes32 internal constant SNAPSHOT_MIGRATION_NOT_ALLOWED = "408"; //Snapshots: Migration only allowed at epoch 0!
+    bytes32 internal constant SNAPSHOT_MIGRATION_INPUT_DATA_MISMATCH = "409"; //Snapshots: Mismatch calldata length!
+    bytes32 internal constant SNAPSHOT_NOT_IN_BUFFER = "410"; //Snapshots: Snapshot no longer in buffer
+    uint256 internal immutable _epochLength;
+
+    uint256 internal immutable _chainId;
+
+    // uint32 internal _epoch;
```
Comment (diligentis) [inline bridge/test/contract-mocks/snapshotRingBuffer/snapshotStorageMock.sol:25]: Delete commented out code.

File: layer1/monitor/events/snapshots.go [hunk 3]
```diff
@@ -12,8 +12,8 @@ import (
 
 // ProcessSnapshotTaken handles receiving snapshots
 func ProcessSnapshotTaken(eth layer1.Client, contracts layer1.AllSmartContracts, logger *logrus.Entry, log types.Log, adminHandler monInterfaces.AdminHandler, taskRequestChan chan<- tasks.TaskRequest) error {
-
```
Comment (diligentis) [inline layer1/monitor/events/snapshots.go:15]: Unless these changes are needed for this PR, better to pull them out into a separate PR that will get much faster review.

File: bridge/test/sharedConstants/4-validators-snapshots-100-Group1.ts [hunk 4]
```diff
@@ -0,0 +1,1010 @@
+import { BigNumber } from "ethers";
+import { ValidatorRawData } from "../ethdkg/setup";
+import { SignedBClaims, Snapshot } from "../setup";
+
+export const validatorsSnapshotsG1: ValidatorRawData[] = [
```
Comment (diligentis) [inline bridge/test/sharedConstants/4-validators-snapshots-100-Group1.ts:5]: Really dumb question: Is it possible to generate this data at test runtime instead of having to check in? With all the magic values here it's entirely inscrutable to anyone without deep understanding.

**Review by diligentis — COMMENTED** [review]
`bridge/test/sharedConstants/constants.ts` is empty and should be removed.

File: bridge/contracts/Snapshots.sol [hunk 5]
```diff
@@ -60,13 +62,19 @@ contract Snapshots is Initializable, SnapshotsStorage, ISnapshots {
...
+        //get the last snapshot
+        Snapshot memory lastSnapshot = getLatestSnapshot();
         require(
-            block.number >= _snapshots[_epoch].committedAt + _minimumIntervalBetweenSnapshots,
+            block.number >= lastSnapshot.committedAt + _minimumIntervalBetweenSnapshots,
             string(abi.encodePacked(SnapshotsErrorCodes.SNAPSHOT_MIN_BLOCKS_INTERVAL_NOT_PASSED))
         );
-
-        uint32 epoch = _epoch + 1;
+        Epoch storage epochReg = _epochReg();
```
Comment (vtleonardo) [inline bridge/contracts/Snapshots.sol:71]: What is epochReg?

File: bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+     * @dev calculates the congruent value for current epoch in respect to the array length
+     * for index to be replaced with most recent epoch
+     * @param epoch_ epoch_ number associated with the snapshot
+     */
+    function _indexFor(SnapshotBuffer storage self_, uint32 epoch_)
+        internal
+        view
+        returns (uint256)
+    {
+        //TODO determine if this is necessary
```
Comment (vtleonardo) [inline bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol:58]: Do we need the require below? The snapshot committed to the contract start at 1.

File: bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+library EpochLib {
+    function set(Epoch storage self_, uint32 value_) internal {
+        self_._value = value_;
+    }
+
+    function get(Epoch storage self_) internal view returns (uint32) {
+        return _max(1, self_._value);
+    }
+
+    //TODO determine if useful
```
Comment (vtleonardo) [inline bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol:73]: We are using it. So it useful I guess.

File: bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    using EpochLib for Epoch;
+
+    // Must be defined in storage contract
+    function _getEpochFromHeight(uint32) internal view virtual returns (uint32);
+
+    // Must be defined in storage contract
+    function _getSnapshots() internal view virtual returns (SnapshotBuffer storage);
+
+    // Must be defined in storage contract
+    function _epochReg() internal view virtual returns (Epoch storage);
```
Comment (vtleonardo) [inline bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol:97]: _epochReg is not a very descriptive name. What is the intent of this variable?

File: bridge/test/sharedConstants/4-validators-snapshots-100-Group1.ts (hunk 4 above)
Comment (vtleonardo) [reply bridge/test/sharedConstants/4-validators-snapshots-100-Group1.ts:5]: Unfortunately no. To generate this data we need a lot of the ETHDKG cryptographic and threshold signature functions that only exist in golang at the moment.

File: bridge/contracts/Snapshots.sol (hunk 5 above)
Comment (z-j-lin) [reply bridge/contracts/Snapshots.sol:71]: sorry about that, I forgot to take out that add but epochReg is the function that returns the struct that holds the epoch. It is implemented in snapshotstorage.sol

File: bridge/contracts/Snapshots.sol [hunk 9]
```diff
@@ -60,13 +62,19 @@ contract Snapshots is Initializable, SnapshotsStorage, ISnapshots {
             IValidatorPool(_validatorPoolAddress()).isConsensusRunning(),
             string(abi.encodePacked(SnapshotsErrorCodes.SNAPSHOT_CONSENSUS_RUNNING))
         );
-
+        //get the last snapshot
```
[the remark this answers is not in the document]
Comment (z-j-lin) [reply bridge/contracts/Snapshots.sol:65]: thank you, will remove

File: layer1/monitor/events/snapshots.go (hunk 3 above)
Comment (vtleonardo) [reply layer1/monitor/events/snapshots.go:15]: Unfortunately, we need this file here, if not sync will break when new nodes join the network. Since we are using the ring buffer now, we don't have a way to get more than 6 snapshots in the past by querying the contract as we used to. Now, the snapshots data (Bclaims) is added to the event and digested with the rest of the events.

**Review by diligentis — APPROVED** [review]
Signing off on what I understand, still blocked on a solidity-reviewer.

**Review by nelsonhp — APPROVED** [review]
